cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,361 @@
+"""
+A general simulator for a birth-death tree simulation process, including 
+fitness. Takes any given distribution on waiting times and fitness 
+mutations.
+"""
+
+import networkx as nx
+import numpy as np
+from queue import PriorityQueue
+
+from typing import Callable, Dict, Generator, Optional, Union
+
+from cassiopeia.data.CassiopeiaTree import CassiopeiaTree
+import cassiopeia.data.utilities as utilities
+from cassiopeia.simulator.TreeSimulator import TreeSimulator, TreeSimulatorError
+
+
+class BirthDeathFitnessSimulator(TreeSimulator):
+
+    def __init__(
+        self,
+        birth_waiting_distribution: Callable[[float], float],
+        initial_birth_scale: float,
+        death_waiting_distribution: Optional[Callable[[], float]] = None,
+        mutation_distribution: Optional[Callable[[], int]] = None,
+        fitness_distribution: Optional[Callable[[], float]] = None,
+        num_extant: Optional[int] = None,
+        experiment_time: Optional[float] = None,
+    ):
+        if num_extant is None and experiment_time is None:
+            raise TreeSimulatorError("Please specify a stopping condition")
+        if num_extant and experiment_time:
+            raise TreeSimulatorError(
+                "Please choose only one stopping condition"
+            )
+        if mutation_distribution is not None and fitness_distribution is None:
+            raise TreeSimulatorError(
+                "Please specify a fitness strength distribution"
+            )
+        if num_extant is not None and num_extant <= 0:
+            raise TreeSimulatorError(
+                "Please specify number of extant lineages greater than 0"
+            )
+        if experiment_time is not None and experiment_time <= 0:
+            raise TreeSimulatorError(
+                "Please specify an experiment time greater than 0"
+            )
+
+        if death_waiting_distribution is None:
+            death_waiting_distribution = lambda: np.inf
+
+        self.birth_waiting_distribution = birth_waiting_distribution
+        self.initial_birth_scale = initial_birth_scale
+        self.death_waiting_distribution = death_waiting_distribution
+        self.mutation_distribution = mutation_distribution
+        self.fitness_distribution = fitness_distribution
+        self.num_extant = num_extant
+        self.experiment_time = experiment_time
+
+    def simulate_tree(
+        self,
+    ) -> CassiopeiaTree:
+        """Simulates trees from a general birth/death process with fitness.
+
+        A forward-time birth/death process is simulated by maintaining a queue
```

Comment:
One comment I have about the documentation is that we should be able to describe the generative model without revealing any implementation details at all. Note that you are revealing implementation details by saying `maintaining a queue`, `At each live currently extant node`, etc. Instead, we can simply say that:
- All cells have the same death waiting time, given by `death_waiting_distribution`.
- Each cell has a specific birth scale parameter, where:
  - The root has birth scale parameter `initial_birth_scale`
  - Whenever a cell divides, it's birth rate scale parameter changes as follows: [...]

And that's the end of if. We don't have to say anything else. All else is an implementation detail, the user doesn't and shouldn't care about it.